### ISSUE TITLE
error while loading shared libraries: /lib64/libc.so.6

### ISSUE THREAD

**Issue by gsal** [containers/buildah#4653]
**Description**

Just getting started with buildah
Reading "Buildah Tutorial 1"
Command does not work

I need to say that I did google this issue and found quite a few threads, many a few years old, but even the one from last years did not seem to have a definite resolution.  They sugges things like re-installing, removing containers directory and starting over; using some restorecon command, etc., etc but nothing solved my condition. 

**Steps to reproduce the issue:**
1.container=$(buildah from fedora)
2.buildah run $container bash

**Describe the results you received:**
```
/usr/bin/bash: error while loading shared libraries: /lib64/libc.so.6: cannot apply additional memory protection after relocation: Permission denied
error while running runtime: exit status 127
```

**Describe the results you expected:**
A bash prompt?

**Output of `rpm -q buildah` or `apt list buildah`:**
```
buildah-1.27.3-1.el9_1.x86_64
```

**Output of `buildah version`:**

```
buildah version  
Version:         1.27.3
Go Version:      go1.18.9
Image Spec:      1.0.2-dev
Runtime Spec:    1.0.2-dev
CNI Spec:        1.0.0
libcni Version:  v1.1.2
image Version:   5.22.1
Git Commit:      
Built:           Mon Jan 23 20:32:26 2023
OS/Arch:         linux/amd64
BuildPlatform:   linux/amd64
```

**Output of `cat /etc/*release`:**

```
NAME="Rocky Linux"
VERSION="9.1 (Blue Onyx)"
ID="rocky"
ID_LIKE="rhel centos fedora"
VERSION_ID="9.1"
PLATFORM_ID="platform:el9"
PRETTY_NAME="Rocky Linux 9.1 (Blue Onyx)"
ANSI_COLOR="0;32"
LOGO="fedora-logo-icon"
CPE_NAME="cpe:/o:rocky:rocky:9::baseos"
HOME_URL="https://rockylinux.org/"
BUG_REPORT_URL="https://bugs.rockylinux.org/"
ROCKY_SUPPORT_PRODUCT="Rocky-Linux-9"
ROCKY_SUPPORT_PRODUCT_VERSION="9.1"
REDHAT_SUPPORT_PRODUCT="Rocky Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="9.1"
Rocky Linux release 9.1 (Blue Onyx)
Rocky Linux release 9.1 (Blue Onyx)
Rocky Linux release 9.1 (Blue Onyx)
Rocky Linux release 9.1 (Blue Onyx)
```

**Output of `uname -a`:**

```
Linux usorla7vp0118x 5.14.0-162.18.1.el9_1.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Mar 1 22:02:24 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# This file is is the configuration file for all tools
# that use the containers/storage library. The storage.conf file
# overrides all other storage.conf files. Container engines using the
# container/storage library do not inherit fields from other storage.conf
# files.
#
#  Note: The storage.conf file overrides other storage.conf files based on this precedence:
#      /usr/containers/storage.conf
#      /etc/containers/storage.conf
#      $HOME/.config/containers/storage.conf
#      $XDG_CONFIG_HOME/containers/storage.conf (If XDG_CONFIG_HOME is set)
# See man 5 containers-storage.conf for more information
# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver, Must be set for proper operation.
driver = "overlay"

# Temporary storage location
runroot = "/run/containers/storage"

# Primary Read/Write location of container storage
# When changing the graphroot location on an SELINUX system, you must
# ensure  the labeling matches the default locations labels with the
# following commands:
# semanage fcontext -a -e /var/lib/containers/storage /NEWSTORAGEPATH
# restorecon -R -v /NEWSTORAGEPATH
graphroot = "/var/lib/containers/storage"


# Storage path for rootless users
#
rootless_storage_path = "$HOME/.local/share/containers/storage"

[storage.options]
# Storage options to be passed to underlying storage drivers

# AdditionalImageStores is used to pass paths to additional Read/Only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Allows specification of how storage is populated when pulling images. This
# option can speed the pulling process of images compressed with format
# zstd:chunked. Containers/storage looks for files within images that are being
# pulled from a container registry that were previously pulled to the host.  It
# can copy or create a hard link to the existing file when it finds them,
# eliminating the need to pull them from the container registry. These options
# can deduplicate pulling of content, disk storage of content and can allow the
# kernel to use less memory when running containers.

# containers/storage supports four keys
#   * enable_partial_images="true" | "false"
#     Tells containers/storage to look for files previously pulled in storage
#     rather then always pulling them from the container registry.
#   * use_hard_links = "false" | "true"
#     Tells containers/storage to use hard links rather then create new files in
#     the image, if an identical file already existed in storage.
#   * ostree_repos = ""
#     Tells containers/storage where an ostree repository exists that might have
#     previously pulled content which can be used when attempting to avoid
#     pulling content from the container registry
pull_options = {enable_partial_images = "false", use_hard_links = "false", ostree_repos=""}

# Remap-UIDs/GIDs is the mapping from UIDs/GIDs as they should appear inside of
# a container, to the UIDs/GIDs as they should appear outside of the container,
# and the length of the range of UIDs/GIDs.  Additional mapped sets can be
# listed and will be heeded by libraries, but there are limits to the number of
# mappings which the kernel will allow when you later attempt to run a
# container.
#
# remap-uids = 0:1668442479:65536
# remap-gids = 0:1668442479:65536

# Remap-User/Group is a user name which can be used to look up one or more UID/GID
# ranges in the /etc/subuid or /etc/subgid file.  Mappings are set up starting
# with an in-container ID of 0 and then a host-level ID taken from the lowest
# range that matches the specified name, and using the length of that range.
# Additional ranges are then assigned, using the ranges which specify the
# lowest host-level IDs first, to the lowest not-yet-mapped in-container ID,
# until all of the entries have been used for maps.
#
# remap-user = "containers"
# remap-group = "containers"

# Root-auto-userns-user is a user name which can be used to look up one or more UID/GID
# ranges in the /etc/subuid and /etc/subgid file.  These ranges will be partitioned
# to containers configured to create automatically a user namespace.  Containers
# configured to automatically create a user namespace can still overlap with containers
# having an explicit mapping set.
# This setting is ignored when running as rootless.
# root-auto-userns-user = "storage"
#
# Auto-userns-min-size is the minimum size for a user namespace created automatically.
# auto-userns-min-size=1024
#
# Auto-userns-max-size is the minimum size for a user namespace created automatically.
# auto-userns-max-size=65536

[storage.options.overlay]
# ignore_chown_errors can be set to allow a non privileged user running with
# a single UID within a user namespace to run containers. The user can pull
# and use any image even those with multiple uids.  Note multiple UIDs will be
# squashed down to the default uid in the container.  These images will have no
# separation between the users in the container. Only supported for the overlay
# and vfs drivers.
#ignore_chown_errors = "false"

# Inodes is used to set a maximum inodes of the container image.
# inodes = ""

# Path to an helper program to use for mounting the file system instead of mounting it
# directly.
#mount_program = "/usr/bin/fuse-overlayfs"

# mountopt specifies comma separated list of extra mount options
mountopt = "nodev,metacopy=on"

# Set to skip a PRIVATE bind mount on the storage home directory.
# skip_mount_home = "false"

# Size is used to set a maximum size of the container image.
# size = ""

# ForceMask specifies the permissions mask that is used for new files and
# directories.
#
# The values "shared" and "private" are accepted.
# Octal permission masks are also accepted.
#
#  "": No value specified.
#     All files/directories, get set with the permissions identified within the
#     image.
#  "private": it is equivalent to 0700.
#     All files/directories get set with 0700 permissions.  The owner has rwx
#     access to the files. No other users on the system can access the files.
#     This setting could be used with networked based homedirs.
#  "shared": it is equivalent to 0755.
#     The owner has rwx access to the files and everyone else can read, access
#     and execute them. This setting is useful for sharing containers storage
#     with other users.  For instance have a storage owned by root but shared
#     to rootless users as an additional store.
#     NOTE:  All files within the image are made readable and executable by any
#     user on the system. Even /etc/shadow within your image is now readable by
#     any user.
#
#   OCTAL: Users can experiment with other OCTAL Permissions.
#
#  Note: The force_mask Flag is an experimental feature, it could change in the
#  future.  When "force_mask" is set the original permission mask is stored in
#  the "user.containers.override_stat" xattr and the "mount_program" option must
#  be specified. Mount programs like "/usr/bin/fuse-overlayfs" present the
#  extended attribute permissions to processes within containers rather then the
#  "force_mask"  permissions.
#
# force_mask = ""

[storage.options.thinpool]
# Storage Options for thinpool

# autoextend_percent determines the amount by which pool needs to be
# grown. This is specified in terms of % of pool size. So a value of 20 means
# that when threshold is hit, pool will be grown by 20% of existing
# pool size.
# autoextend_percent = "20"

# autoextend_threshold determines the pool extension threshold in terms
# of percentage of pool size. For example, if threshold is 60, that means when
# pool is 60% full, threshold has been hit.
# autoextend_threshold = "80"

# basesize specifies the size to use when creating the base device, which
# limits the size of images and containers.
# basesize = "10G"

# blocksize specifies a custom blocksize to use for the thin pool.
# blocksize="64k"

# directlvm_device specifies a custom block storage device to use for the
# thin pool. Required if you setup devicemapper.
# directlvm_device = ""

# directlvm_device_force wipes device even if device already has a filesystem.
# directlvm_device_force = "True"

# fs specifies the filesystem type to use for the base device.
# fs="xfs"

# log_level sets the log level of devicemapper.
# 0: LogLevelSuppress 0 (Default)
# 2: LogLevelFatal
# 3: LogLevelErr
# 4: LogLevelWarn
# 5: LogLevelNotice
# 6: LogLevelInfo
# 7: LogLevelDebug
# log_level = "7"

# min_free_space specifies the min free space percent in a thin pool require for
# new device creation to succeed. Valid values are from 0% - 99%.
# Value 0% disables
# min_free_space = "10%"

# mkfsarg specifies extra mkfs arguments to be used when creating the base
# device.
# mkfsarg = ""

# metadata_size is used to set the `pvcreate --metadatasize` options when
# creating thin devices. Default is 128k
# metadata_size = ""

# Size is used to set a maximum size of the container image.
# size = ""

# use_deferred_removal marks devicemapper block device for deferred removal.
# If the thinpool is in use when the driver attempts to remove it, the driver
# tells the kernel to remove it as soon as possible. Note this does not free
# up the disk space, use deferred deletion to fully remove the thinpool.
# use_deferred_removal = "True"

# use_deferred_deletion marks thinpool device for deferred deletion.
# If the device is busy when the driver attempts to delete it, the driver
# will attempt to delete device every 30 seconds until successful.
# If the program using the driver exits, the driver will continue attempting
# to cleanup the next time the driver is used. Deferred deletion permanently
# deletes the device and all data stored in device will be lost.
# use_deferred_deletion = "True"

# xfs_nospace_max_retries specifies the maximum number of retries XFS should
# attempt to complete IO when ENOSPC (no space) error is returned by
# underlying storage device.
# xfs_nospace_max_retries = "0"

```

**Comment by flouthoc** [comment]
Could you please try again with a more recent version ? Since I am unable to reproduce this from the upstream version.

```console
$ container=$(./buildah from fedora)
Resolved "fedora" as an alias (/etc/containers/registries.conf.d/000-shortnames.conf)
Trying to pull registry.fedoraproject.org/fedora:latest...
Getting image source signatures
Copying blob 6f70e28bcc6a done  
Copying config 38b8a6c320 done  
Writing manifest to image destination
Storing signatures
[fl@fedora bin]$ buildah run $container bash
[root@4fa5eb036eeb /]# exit
exit
[fl@fedora bin]$ 
```

**Comment by gsal** [comment]
Hi, thanks for your reply. 

I just realized that nowhere in the title nor posting do I mention that my troubles are in root-less mode; the only hint could be the uncommented `rootless_storage_path` line in the `storage.conf` file.

Over the weekend, I continued to browse through various articles and tutorials until I found the needed conditions and needed commands to be able to build a root-less image.

I found that there are two key pieces of information that are not mentioned whatsoever (that I could tell) in most tutorials, including this one: https://github.com/containers/buildah/blob/main/docs/tutorials/01-intro.md

They are:

- The user's home directory cannot just be NFS mounted, it seems that it must exist locally in the same machine where we are executing the buildah command
- Previous to start issuing the typical buildah commands, when building a root-less image, the user needs to first issue the `buildah unshare` command to enter some kind of shell withih the user namespace; and, aftewards, the user must issue `exit` to exit such shell.

Maybe, just maybe, the above is obvious to some IT person or an experience DevOps; but, for as long as we have tutorials, it should be assume the reader does not know much.

**Comment by gsal** [comment]
Hi, it's me, again. 

So, as mentioned in my previous post, during the build of the image, I was able to get pass the `error while loading...libc.so.6` by using `buildah unshare`; but, now, I seem to run into the same trouble while attempting to run the container with podman.

Straight up:  It first says that I need a user session
```
> podman run -it --rm myimage
WARN[0000] The cgroupv2 manager is set to systemd but there is no systemd user session available 
WARN[0000] For using systemd, you may need to login using an user session 
WARN[0000] Alternatively, you can enable lingering with: `loginctl enable-linger 1001` (possibly as root) 
WARN[0000] Falling back to --cgroup-manager=cgroupfs    
WARN[0000] The cgroupv2 manager is set to systemd but there is no systemd user session available 
WARN[0000] For using systemd, you may need to login using an user session 
WARN[0000] Alternatively, you can enable lingering with: `loginctl enable-linger 1001` (possibly as root) 
WARN[0000] Falling back to --cgroup-manager=cgroupfs    
/bin/sh: error while loading shared libraries: libc.so.6: cannot change memory protections
```

First issuing `buildah unshare`: 

```
> buildah unshare
> podman run -it --rm myimage
WARN[0000] The cgroupv2 manager is set to systemd but there is no systemd user session available 
WARN[0000] For using systemd, you may need to login using an user session 
WARN[0000] Alternatively, you can enable lingering with: `loginctl enable-linger 1001` (possibly as root) 
WARN[0000] Falling back to --cgroup-manager=cgroupfs    
/bin/sh: error while loading shared libraries: libc.so.6: cannot change memory protections
```
Only one warning instead of two; but no cigar.

And, of course, I don't know what to make out of the above warnings.
Hinst, please?
How can I run my own image?

**Comment by gsal** [comment]
Here is my buildah session, running in Linux/Rocky9: 

```
> export cwd=${PWD}
> export wcont=$(buildah from scratch)
> buildah unshare
> mpoint=$(buildah mount $wcont)
> dnf install -y \
    --installroot ${mpoint} \
    --releasever 9.1 bash coreutils which tree gcc gfortran \
    --setopt install_weak_deps=false
.
.
.
Complete!
> buildah run $wcont bash
bash-5.1#
bash-5.1# exit
exit
> buildah commit $wcont myimage
> buildah unmount $wcont
> exit # to exit buildah unshare
> buildah rm $wcont
> 
> # try to run image
>
> buildah images
REPOSITORY          TAG      IMAGE ID       CREATED         SIZE
localhost/myimage   latest   311d06b13768   2 minutes ago   348 MB
>
> podman images 
WARN[0000] The cgroupv2 manager is set to systemd but there is no systemd user session available 
WARN[0000] For using systemd, you may need to login using an user session 
WARN[0000] Alternatively, you can enable lingering with: `loginctl enable-linger 1001` (possibly as root) 
WARN[0000] Falling back to --cgroup-manager=cgroupfs    
WARN[0000] The cgroupv2 manager is set to systemd but there is no systemd user session available 
WARN[0000] For using systemd, you may need to login using an user session 
WARN[0000] Alternatively, you can enable lingering with: `loginctl enable-linger 1001` (possibly as root) 
WARN[0000] Falling back to --cgroup-manager=cgroupfs    
REPOSITORY         TAG         IMAGE ID      CREATED        SIZE
localhost/myimage  latest      311d06b13768  2 minutes ago  348 MB
>
> podman run -it --rm myimage
WARN[0000] The cgroupv2 manager is set to systemd but there is no systemd user session available 
WARN[0000] For using systemd, you may need to login using an user session 
WARN[0000] Alternatively, you can enable lingering with: `loginctl enable-linger 1001` (possibly as root) 
WARN[0000] Falling back to --cgroup-manager=cgroupfs    
WARN[0000] The cgroupv2 manager is set to systemd but there is no systemd user session available 
WARN[0000] For using systemd, you may need to login using an user session 
WARN[0000] Alternatively, you can enable lingering with: `loginctl enable-linger 1001` (possibly as root) 
WARN[0000] Falling back to --cgroup-manager=cgroupfs    
/bin/sh: error while loading shared libraries: libc.so.6: cannot change memory protections
```

**Comment by rhatdan** [comment]
This looks like an SELinux issue.  Most likely container-selinux failed to install.

yum -y reinstall container-selinux

And see if it throws an error.

There is a problem on RHEL9 right now, you might need to downgrade your container-selinux or see if there is an updated version of libsepol or container-selinux available.

**Comment by gsal** [comment]
Thanks for the quick reply.

Reminder: 
-While using buildah, I am a regular user without root privileges
-While using yum in the host, I do have the ability to become root. 

I re-installed: 
```
.
.
.
Reinstalled:
  container-selinux-3:2.189.0-1.el9.noarch                                                       

Complete!
```

Tried creating an image from fedora, for example....still does not work

```
1203> buildah unshare 
1204> export wcont=$(buildah from fedora:latest)
Resolved "fedora" as an alias (/etc/containers/registries.conf.d/000-shortnames.conf)
Trying to pull registry.fedoraproject.org/fedora:latest...
Getting image source signatures
Copying blob 6f70e28bcc6a done  
Copying config 38b8a6c320 done  
Writing manifest to image destination
Storing signatures
1205> buildah run $wcont bash 
/usr/bin/bash: error while loading shared libraries: /lib64/libc.so.6: cannot apply additional memory protection after relocation: Permission denied
error while running runtime: exit status 127
1206> 
```

Looked up libsepol: 
```
> yum info libsepol
Last metadata expiration check: 2:26:30 ago on Wed 15 Mar 2023 04:00:00 PM EDT.
Installed Packages
Name         : libsepol
Version      : 3.4
Release      : 1.1.el9
Architecture : x86_64
Size         : 771 k
Source       : libsepol-3.4-1.1.el9.src.rpm
Repository   : @System
From repo    : anaconda
Summary      : SELinux binary policy manipulation library
URL          : https://github.com/SELinuxProject/selinux/wiki
License      : LGPLv2+
.
.
.
```
Additional info.
I did a yum search on container, found 5 additional packages, 3 already installed, 2 not yet: 

Installed: 
-container-selinux.noarch 
-containers-common.x86_64 
-conmon.x86_64            

Not yet:
-container-tools.noarch    
-systemd-container.x86_64

**Comment by gsal** [comment]
Thanks for that command, I did try that one and a few others, repeatedly, in different sequence, etc.

Not in any particular order, I tried:
- podman system reset
- restorecon -R -v $HOME
- semanage fcontext -a -e /var/lib/containers /data/containers
- restorecon -R -vv /data/containers

I also took root-owned storage.conf and placed a copy of my own under ~/.config/containers.
Also set: graphroot = "$HOME/.local/share/containers/storage"

What finally seemed to have done the trick was changing my storage.conf and set: 
runroot = "/run/user/<UID>/containers/storage"

Unfortunately, I still don't know the minimal recipe, but at least I can use buildah to build an image and podman to run it. We shall see what problems I run into when I ask somebody else to try and do the same. 

Until the next one.
